### PR TITLE
320 authentication client for frontend

### PR DESCRIPTION
**Pull request by absurd-turtle**
## Basics

- [x] Details of what you changed are in commit messages.
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildserver is happy.
- [x] The PR is rebased with current master.

## Checklist

- [ ] I added unit tests for my code
- [x] I fully described what my PR does in the documentation
      (not in the PR description)
- [x] I fixed all affected documentation
- [x] I fixed all affected decisions
- [x] I added code comments, logging, and assertions as appropriate
- [x] I translated all strings visible to the user
- [x] I mentioned [every code or binary](/.reuse/dep5) not directly written or done by me in [reuse syntax](https://reuse.software/)

## Review

- [ ] I've tested the code
- [ ] I've read through the whole code
- [ ] Documentation is conforming to [our Documentation Guidelines](/doc/documentation.md)
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to our Coding Guidelines
- [ ] Code is consistent to [our Design Decisions](/doc/decisions)

### CODE REVIEW

**Comment by markus2330** [comment]
Target date to merge this PR: Tuesday 14:00.

File: backend/.env.sample [hunk 1]
```diff
@@ -2,3 +2,7 @@
 DATABASE_URL=postgres://permaplant:permaplant@127.0.0.1/permaplant
 BIND_ADDRESS_HOST=127.0.0.1
 BIND_ADDRESS_PORT=8080
+
+# OAuth2 (other URLs will be fetched from this URL)
+AUTH_DISCOVERY_URI=http://localhost:8081/realms/PermaplanT/.well-known/openid-configuration
+AUTH_CLIENT_ID=PermaplanT
```
Comment (markus2330) [inline backend/.env.sample:8]: Why would we want to configure this?

File: backend/README.md [hunk 2]
```diff
@@ -13,6 +16,8 @@
 - `DATABASE_URL` is the Connection [URI](https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNSTRING) to your PostgreSQL database
 - `BIND_ADDRESS_HOST` defines the host on which the server will run on
 - `BIND_ADDRESS_PORT` defines the port on which the server will run on
+- `AUTH_DISCOVERY_URI` the .well-known endpoint of the auth server (see [RFC 8414](https://www.rfc-editor.org/rfc/rfc8414.html#section-2) for more detail)
+- `AUTH_CLIENT_ID` the client id the frontend should use to log in
```
Comment (markus2330) [inline backend/README.md:20]: Let us hard-code this, it also makes the description in "Setup Keycloak" again correct.

File: backend/README.md [hunk 3]
```diff
@@ -45,19 +50,44 @@ LC_ALL=C diesel migration run
...
+Try to execute a request (it should return error 401).  
+Click `Authorize`.
 
-Example requests:
-
-- `curl localhost:8080/api/plants`
+> Use the `authorizationCode` auth flow.  
+> Enter client_id `swagger-ui` (client_secret is empty) and click `Authorize`.  
+> Enter user credentials (username: `test`, password: `test`).  
+> You should now be able to execute a request in Swagger.
```
Comment (markus2330) [inline backend/README.md:90]: Why would this be a note? It is an essential part to test server using Swagger

```suggestion
- Use the `authorizationCode` auth flow.
- Enter client_id `swagger-ui` (client_secret is empty) and click `Authorize`.
- Enter user credentials (username: `test`, password: `test`).
- You should now be able to execute a request in Swagger.
```

File: backend/README.md [hunk 4]
```diff
@@ -45,19 +50,44 @@ LC_ALL=C diesel migration run
...
 cargo run
 ```
 
-## Usage
+### Test server using Swagger
 
-Now the server is running and will start listening at <http://localhost:8080/> (or whichever port you specified in `.env`).
+Go to <http://localhost:8080/doc/api/swagger/ui/>.  
+Try to execute a request (it should return error 401).  
+Click `Authorize`.
```
Comment (markus2330) [inline backend/README.md:85]:
```suggestion
- Go to <http://localhost:8080/doc/api/swagger/ui/>.
  If you try to execute a request now it should return error 401.
- Click `Authorize`.
```

File: devel/keycloak/docker-compose.yml
Comment (markus2330) [inline devel/keycloak/docker-compose.yml:1]: File doesn't seem to be used nor documented. If you add a README.md what it is for, you can move it to `doc/setups` or similar.

File: devel/nextcloud/docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    build:
+      context: .
+      dockerfile: ../nginx/Dockerfile
+    ports:
+      - "443:443"
+      - "80:80"
+    restart: always
+
+  db:
+    image: mariadb:10.6
```
Comment (markus2330) [inline devel/nextcloud/docker-compose.yml:18]: Wouldn't it make more sense to use PostgreSQL everywhere?

File: frontend/src/App.tsx [hunk 6]
```diff
@@ -1,37 +1,57 @@
...
+    switch (auth.activeNavigator) {
+      case 'signinSilent':
+        toast('Signing you in...');
+        break;
+      case 'signoutRedirect':
+        toast(`Signing you out...`);
+    }
+    console.log('authenticated: ', auth.isAuthenticated);
+    if (auth.isAuthenticated) {
+      toast(`Hello ${auth.user?.profile.sub}`);
```
Comment (markus2330) [inline frontend/src/App.tsx:39]: This comes quite often and prints a long hexadecimal number in my case. It is also missing translation.

File: frontend/src/App.tsx [hunk 7]
```diff
@@ -1,37 +1,57 @@
...
+
+  // authentication notifications
+  useEffect(() => {
+    console.log('auth loading: ', auth.isLoading);
+    if (auth.error) {
+      toast(`Oops... ${auth.error.message}`);
+    }
+    switch (auth.activeNavigator) {
+      case 'signinSilent':
+        toast('Signing you in...');
```
Comment (markus2330) [inline frontend/src/App.tsx:32]: Please translate all strings.

File: frontend/src/features/auth/components/LoginButton.tsx [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+};
+
+export const LoginButton = () => {
+  const auth = useAuth();
+
+  if (auth.isAuthenticated) {
+    return (
+      <div>
+        <SimpleButton className="pl-6 pr-0" onClick={() => void auth.removeUser()}>
+          Log out
```
Comment (markus2330) [inline frontend/src/features/auth/components/LoginButton.tsx:16]: Please translate all strings.

File: frontend/src/features/landing_page/api/getImages.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
+import { createNextcloudAPI } from '@/config/axios';
+
+export const getPhotos = async (): Promise<Array<string>> => {
```
Comment (markus2330) [inline frontend/src/features/landing_page/api/getImages.ts:3]: Great to see that but it would probably be better to have this in a different PR. It definitely needs some documentation what it is for.

File: frontend/src/routes/index.ts [hunk 10]
```diff
@@ -43,6 +44,11 @@ const routes: Routes = {
     path: '/maps',
     title: 'Map Overview',
   },
+  [Pages.Login]: {
+    component: Login,
+    path: '/login',
```
Comment (markus2330) [inline frontend/src/routes/index.ts:49]: Why make it configurable if it is hard-coded after all. VITE_REDIRECT_URI can probably be removed, too. @Bushuo already worked on an issue where such a useless config was removed.

**Review by markus2330 — CHANGES_REQUESTED** [review]
Great job, everything worked for me. Only some simplifications and documentation is needed.

**Comment by markus2330** [comment]
One quite nasty bug: when the frontend cannot reach keycloak or backend anymore, the whole landing page gets replaced by:

image

Expected behavior:
User simply gets logged out by timeout and still sees the landing page.

Can also be fixed by follow-up issue, but please create the follow-up issues first.

File: backend/.env.sample (hunk 1 above)
Comment (kitzbergerg) [reply backend/.env.sample:8]: There are different clients configured for different instances e.g. https://www.permaplant.net or https://pr.permaplant.net/.
Those need different CLIENT_IDs.

File: backend/README.md (hunk 2 above)
Comment (kitzbergerg) [reply backend/README.md:20]: see comment above

File: backend/README.md (hunk 3 above)
Comment (kitzbergerg) [reply backend/README.md:90]: No idea how that happened, but it wasn't intended.

File: frontend/.env.sample [hunk 11]
```diff
@@ -1,4 +1,14 @@
...
+# Env variables must have the prefix VITE_ for Vite to include them.
+# More information about env variables with Vite can be found here: https://vitejs.dev/guide/env-and-mode.html
 
 VITE_BASE_API_URL=http://localhost:8080
+
+VITE_NEXTCLOUD_URI=http://localhost:8081
+
+# authentication
+# The URL of the OIDC/OAuth2 provider.
+# VITE_AUTHORITY=http://0.0.0.0:8081/realms/PermaplanT
```
[the remark this answers is not in the document]
Comment (kitzbergerg) [reply frontend/.env.sample:10]: @absurd-turtle This whole section should probably be removed from the frontend it isn't necessary any more.

File: frontend/.env.sample [hunk 12]
```diff
@@ -1,4 +1,14 @@
...
+# client application's identifier as registered with the OIDC/OAuth2 provider.
+# VITE_CLIENT_ID=PermaplanT
```
[the remark this answers is not in the document]
Comment (kitzbergerg) [reply frontend/.env.sample:12]: env vars should be removed entirely from frontend. See above

File: frontend/.env.sample [hunk 13]
```diff
@@ -1,4 +1,14 @@
...
+# The redirect URI of your client application to receive a response from the OIDC/OAuth2 provider.
+# VITE_REDIRECT_URI=http://localhost:5173/login
```
[the remark this answers is not in the document]
Comment (kitzbergerg) [reply frontend/.env.sample:14]: env vars should be removed entirely from frontend. See above

File: devel/keycloak/docker-compose.yml
Comment (absurd-turtle) [reply devel/keycloak/docker-compose.yml:1]: I used it to run my local instance of keycloak for testing and development purposes.
I don't know if we will need it later on or if we want to test against the production instance or some dedicated online development instance. 
The same goes for Nextcloud.
If we want to have some local dev containers for these services I suggest we improve on these docker files.
As of now keycloak has to be configured after creating the instance. This could be improved by having the needed settings in a build script so not every dev will have to concern themselves with setting up keycloak.
Maybe we can also reuse what @lukashartl wrote for the production environment.

File: frontend/.env.sample (hunk 13 above)
[the remark this answers is not in the document]
Comment (markus2330) [reply frontend/.env.sample:14]: perfect!

File: frontend/src/features/landing_page/api/getImages.ts (hunk 9 above)
Comment (absurd-turtle) [reply frontend/src/features/landing_page/api/getImages.ts:3]: Yes, as mentioned in some other PR. The purpose was for testing requests with credentials against the NextcloudAPI.
I can add a comment mentioning this, so we can improve on it in a dedicated issue.

Comment (markus2330) [reply frontend/src/features/landing_page/api/getImages.ts:3]: Yes, write that at the top of the file. Then everything is clear.

File: frontend/src/routes/index.ts (hunk 10 above)
Comment (absurd-turtle) [reply frontend/src/routes/index.ts:49]: Yes, correct.
In fact the whole route, component and env variable is rendered useless with the recent changes.
I will remove it.

**Comment by absurd-turtle** [comment]
> One quite nasty bug: when the frontend cannot reach something (keycloak or backend?) anymore, the whole landing page gets replaced by:
> 
> image
> 
> Expected behavior: User simply gets logged out by timeout and still sees the landing page.
> 
> Can also be fixed by follow-up issue, but please create the follow-up issues first.

This happens when the backend isn't reachable.
Good point though. There is no reason why the landing page should not be displayed. I will look into it.

File: frontend/src/features/landing_page/api/getImages.ts (hunk 9 above)
Comment (absurd-turtle) [reply frontend/src/features/landing_page/api/getImages.ts:3]: There is probably a better location for it though.

Comment (absurd-turtle) [reply frontend/src/features/landing_page/api/getImages.ts]: I created a separate issue and branch for this: #364 
and moved the functions there

File: backend/.env.sample (hunk 1 above)
Comment (kitzbergerg) [reply backend/.env.sample:8]: I would ask @lukashartl to add that there then. Maybe even write in the doc how the client_id will be called per deployment.

@markus2330 Another option would be to hardcode the client_id (although I don't really like hardcoding values like that) and instead use different realms.
The current solution would mean we can/have to use the same users for https://dev.permaplant.net, https://pr.permaplant.net and https://www.permaplant.net. With different realms users are separate and you could for example only allow us developers on https://pr.permaplant.net. 

Again I would refer to @lukashartl for that as he is the one setting up Keycloak.

File: devel/keycloak/docker-compose.yml
Comment (absurd-turtle) [reply devel/keycloak/docker-compose.yml:1]: The only reason was because the documentation in the backend simply did not exist at the time and I didn't really know where to put the docker-compose file.
In general I prefer using docker-compose as the commands can get quite long when adding volumes and such.

File: backend/.env.sample (hunk 1 above)
Comment (markus2330) [reply backend/.env.sample:8]: Yes, I have to agree it would be better to have the same users on all three instances. But what is the purpose of the client_id then? Is it possible to have a test user for only one of the instances?

Comment (kitzbergerg) [reply backend/.env.sample:8]: The have different origins. In Keycloak you have to define the root URL and also where you are allowed to redirect users to for each client.
E.g. https://dev.permaplant.net is a different root URL than https://pr.permaplant.net. So if you try to log in via a client that has https://dev.permaplant.net as root URL, but you make that call from https://pr.permaplant.net (you try to reuse a client) Keycloak wouldn't allow that. I think it would technically be possible to define multiple allowed redirect URLs, but that seems very hacky.

**Comment by absurd-turtle** [comment]
> > One quite nasty bug: when the frontend cannot reach something (keycloak or backend?) anymore, the whole landing page gets replaced by:
> > image
> > Expected behavior: User simply gets logged out by timeout and still sees the landing page.
> > Can also be fixed by follow-up issue, but please create the follow-up issues first.
> 
> This happens when the backend isn't reachable. Good point though. There is no reason why the landing page should not be displayed. I will look into it.

Unfortunately there is no easy fix for this.
The AuthProvider expects the correct configuration. We can delay the initialization of the AuthProvider and wait for the configuration from the backend and then intitialize it.
If the backend is down the AuthProvider never receives the config and we can return an error message.
This was our first approach. This leads to the problem that nothing except an error message is displayed because the AuthProvider can't be initialized.

We tried to fix this by passing a default config.
Then wait for the config from the backend (until it is online again) and then rerendering the AuthProvider component.
This can achieved by providing a key prop:

`const config = data ?? defaultConfig;`
`return <AuthProvider key={config.authority} {...config}>{children}</AuthProvider>;`

Now the problem is that once the configuration has changed and the AuthProvider is reloaded the login does not work anymore.

**Comment by markus2330** [comment]
@Bushuo @badnames @buenaflor can you take a look here?

File: devel/keycloak/docker-compose.yml [hunk 14]
```diff
@@ -0,0 +1,22 @@
+version: "2"
```
Comment (markus2330) [inline devel/keycloak/docker-compose.yml:1]: Please move to `doc/setups`, `devel` could be anything.

**Review by Bushuo — APPROVED** [review]

**Review by markus2330 — CHANGES_REQUESTED** [review]
Last fix: please also add how to use public instance.

**Comment by markus2330** [comment]
jenkins build please

File: backend/.env.sample (hunk 1 above)
Comment (markus2330) [reply backend/.env.sample:8]: The documentation also lacks which are valid client IDs for auth.permaplant.net (PermaplanT didn't work for me)

Changelog should contain both that:

- now the config gets fetched from backend
- that auth is active

File: backend/README.md [hunk 15]
```diff
@@ -52,8 +52,19 @@ cargo install typeshare-cli
 
 6. Start Keycloak
 
+via docker run:
```
[the remark this answers is not in the document]
Comment (lukashartl) [reply backend/README.md]: I have created the client as requested.

* Client ID = `localhost`
* Root URL = `http://localhost:5173/`
* Valid Redirect URLs = `/*`
* Web Origins = `+`

@absurd-turtle , @markus2330 could someone pls verify whether this works? 😄

[the remark this answers is not in the document]
Comment (markus2330) [reply backend/README.md]: Thank you, following .env works flawlessly for me (including self-registration):

```
DATABASE_URL=postgres://permaplant:permaplant@127.0.0.1/permaplant
BIND_ADDRESS_HOST=localhost
BIND_ADDRESS_PORT=8080
AUTH_DISCOVERY_URI=http://auth.permaplant.net/realms/PermaplanT/.well-known/openid-configuration
AUTH_CLIENT_ID=localhost
```

@absurd-turtle or @kitzbergerg please do the last updates, so that we can finally merge this.

**Review by lukashartl — APPROVED** [review]

File: backend/.env.sample (hunk 1 above)
Comment (kitzbergerg) [reply backend/.env.sample:8]: The configuration was still set up for the local Keycloak instance and there exists a client ID PermplanT as described in the setup steps.
I changed the sample to point to the dev instance on auth.permaplant

**Review by kitzbergerg — APPROVED** [review]

File: backend/.env.sample (hunk 1 above)
Comment (kitzbergerg) [reply backend/.env.sample]: The client IDs needed for development are in the documentation (in the setup steps).
With changelog you mean add it in the next meeting notes? Or do we have an actual changelog? If so I couldn't find it.

**Comment by absurd-turtle** [comment]
> > > One quite nasty bug: when the frontend cannot reach something (keycloak or backend?) anymore, the whole landing page gets replaced by:
> > > image
> > > Expected behavior: User simply gets logged out by timeout and still sees the landing page.
> > > Can also be fixed by follow-up issue, but please create the follow-up issues first.
> > 
> > 
> > This happens when the backend isn't reachable. Good point though. There is no reason why the landing page should not be displayed. I will look into it.
> 
> Unfortunately there is no easy fix for this. The AuthProvider expects the correct configuration. We can delay the initialization of the AuthProvider and wait for the configuration from the backend and then intitialize it. If the backend is down the AuthProvider never receives the config and we can return an error message. This was our first approach. This leads to the problem that nothing except an error message is displayed because the AuthProvider can't be initialized.
> 
> We tried to fix this by passing a default config. Then wait for the config from the backend (until it is online again) and then rerendering the AuthProvider component. This can achieved by providing a key prop:
> 
> `const config = data ?? defaultConfig;` `return <AuthProvider key={config.authority} {...config}>{children}</AuthProvider>;`
> 
> Now the problem is that once the configuration has changed and the AuthProvider is reloaded the login does not work anymore.
> 
> I have created a seperate issue for this here: [#365](https://github.com/ElektraInitiative/PermaplanT/issues/365)

This issue is now fixed in [#365](https://github.com/ElektraInitiative/PermaplanT/issues/365)

**Review by markus2330 — APPROVED** [review]
@kitzbergerg mdbook checks unfortunately still fails.

Other than that we can merge from my point of view.

**Comment by kitzbergerg** [comment]
@markus2330 should be done

**Comment by markus2330** [comment]
pr.permaplant.net should be tested too

**Comment by markus2330** [comment]
Finally done! Great job!

File: backend/.env.sample (hunk 1 above)
Comment (markus2330) [reply backend/.env.sample]: Yes, the Changelog in each meeting note is our actual changelog. For this week #377
